### ISSUE TITLE
tensordot vs tensor_dot vs contract vs inner 

### ISSUE THREAD

**Issue by merajhashemi** [tensorly/tensorly#250]
All the backends Tensorly supports have a built-in `tensordot` functions which implements a general tensor contraction (e.g. [numpy](https://numpy.org/doc/stable/reference/generated/numpy.tensordot.html)). However, in Tensorly `tl.tenalg.tensor_dot` is a generalized outer product. This may cause trouble for new users.

Suggestions:
1. Adding `tl.tensordot` as a wrapper for the backends `tensordot` function.
2. Deprecating `tl.tenalg.tensor_dot` in favour of an improved `tl.tenalg.outer` function which supports tensors.
3. Replacing `tl.tenalg.contract` and `tl.tenalg.inner` implementations with `tl.tensordot`, or even removing them as both functions can be replicated with one line using the new `tl.tensordot`:?

**Comment by JeanKossaifi** [comment]
I agree, it's great to see the array libraries natively supporting more tensor operations! I have found the API of `tensordot` a little misleading/confusing for the user, as it can do tensor contractions *and* actual tensor product. However I do think it's good to be consistent with NumPy API which users might be familiar with. 

1. Agreed
2. Also agreed
3. I think it can be useful to have for some use cases, even though they are indeed trivial to rewrite using `tensordot`. We can at least implement them for now and deprecate them in favor of `tensordot`? 

The only consideration is the `"einsum"` tenalg backend for dispatching the operations to an einsum interface: `tensordot` will be slightly trickier to write with all the use-cases.

**Comment by JeanKossaifi** [comment]
I added tensordot in https://github.com/tensorly/tensorly/commit/e27127f22572954f17cb17e516828b7d93b9887c, 2 and 3 remaining.
